File: src/c4/inter/Arith.java
package inter;
import treewalker.TreeWalker;
import code.DreiAdrCode;
import code.Arith2OpCode;

import lexer.*;

/*
 * Arith ist eine Unterklasse von Op und beschreibt arithmetische Ausdrücke. 
 * In den Instanzenvariablen expr1 und expr2 werden die beiden Operanden 
 * abgelegt
 */

public class Arith extends Op {
	Expr expr1, expr2;

	public Arith(Token tok, Expr x1, Expr x2) {
		super(tok);
		expr1 = x1;
		expr2 = x2;
	}
	
	// für TransformWalker:
	
	public Arith(Token tok, Expr x1, Expr x2, Type p) {
		super(tok);
		expr1 = x1;
		expr2 = x2;
		type = p;	
	}

	
	public Expr getExpr1() {
		return expr1;
	}

	public Expr getExpr2() {
		return expr2;
	}
	
	public void setExpr1(Expr expr1) {
		this.expr1 = expr1;
	}

	public void setExpr2(Expr expr2) {
		this.expr2 = expr2;
	}

	// für die Drei-Adress-Code Erzeugung
	public DreiAdrCode codeForValueTo(Id id) {
		return (new Arith2OpCode(id, this));
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkArithNode(this, arg);
	}

}

File: src/c4/inter/Arith.rs
use super::{
    DreiAdrCode, Expr, ExprData, ExpressionNode, Id, Node, NodeData, OpData, Token, TreeWalker,
    Type, Walk,
};

/// Binary arithmetic expression.
#[derive(Clone, Debug, PartialEq)]
pub struct Arith {
    op: OpData,
    expr1: Expr,
    expr2: Expr,
}

impl Arith {
    pub fn new(token: Token, expr1: Expr, expr2: Expr) -> Self {
        Self {
            op: OpData::new(token),
            expr1,
            expr2,
        }
    }

    pub fn with_type(token: Token, expr1: Expr, expr2: Expr, ty: Type) -> Self {
        let mut arith = Self::new(token, expr1, expr2);
        arith.set_type(ty);
        arith
    }

    pub fn get_expr1(&self) -> &Expr {
        &self.expr1
    }
    pub fn get_expr2(&self) -> &Expr {
        &self.expr2
    }
    pub fn set_expr1(&mut self, expr: Expr) {
        self.expr1 = expr;
    }
    pub fn set_expr2(&mut self, expr: Expr) {
        self.expr2 = expr;
    }

    pub fn code_for_value_to(&self, target: Id) -> DreiAdrCode {
        let left = self
            .expr1
            .as_singleton()
            .expect("left operand must be lowered to a Singleton");
        let right = self
            .expr2
            .as_singleton()
            .expect("right operand must be lowered to a Singleton");
        DreiAdrCode::Binary {
            target,
            left,
            operator: self.get_op().clone(),
            right,
        }
    }
}

impl Node for Arith {
    fn node_data(&self) -> &NodeData {
        &self.op.expr.node
    }
}

impl ExpressionNode for Arith {
    fn expr_data(&self) -> &ExprData {
        &self.op.expr
    }
    fn expr_data_mut(&mut self) -> &mut ExprData {
        &mut self.op.expr
    }
}

impl<P, R> Walk<P, R> for Arith {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_arith_node(self, arg)
    }
}
